### PR TITLE
Sped up page load time with data files optimization at build time

### PR DESCRIPTION
**Pull request by Sp3EdeR**
Notes for reviewer:
* https://kmlviewer.nsspot.net/ can be used to open KML files online.
* https://sp3eder.github.io/huroutes/linkmaker.html can be used to decode location marker URLs.

### CODE REVIEW

File: res/huroutes.js [hunk 1]
```diff
@@ -493,7 +493,13 @@ function addRoute(data)
...
     layer.routeId = routeId;
     // Load the KML file data into the Leaflet layer
-    omnivore.kml(data.kml, null, layer).addTo(map);
+    if (data.kml.substr(-3) === '.kml')
+        omnivore.kml(data.kml, null, layer).addTo(map);
+    else
+        tmout = setTimeout(() => {
+            omnivore.kml.parse(data.kml, null, layer).addTo(map);
+            clearTimeout(tmout);
+        }, 0);
```
Comment (nbali) [inline res/huroutes.js:502]: async explanation

File: .github/scripts/build-tools/compact_data.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        if isinstance(data, list):
+            for value in data:
+                yield from self._routes(value)
+        elif isinstance(data, dict):
+            for key, value in data.items():
+                if key in self._routeKeys:
+                    yield data
+                    break
+                else:
+                    yield from self._routes(value)
```
Comment (nbali) [inline .github/scripts/build-tools/compact_data.py:25]: this loop won't necessarily do what u expect imo, at least I assume your goal is to return the whole "dict" if we have the `_routeKeys` present, and only try to use the "value" if both of those properties are missing

the problem is that we might encounter the first property that isn't md, or kml, so we yield with the else branch, but in reality we might have those properties

shouldn't this be:
```python
if (data.keys() & self._routeKeys)):
  yield data
else:
  yield from self._routes(value)
```

**Review by nbali — CHANGES_REQUESTED** [review]

File: .github/scripts/build-tools/compact_data.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        if not route[type].endswith('.' + type):
+            return
+        try:
+            if 'id' not in route:
+                route['id'] = re.match('data\/([\w-]+).' + type, route[type])[1]
+            trackPath = os.path.join(self._dataDir, route[type])
+            with open(trackPath, encoding='utf-8') as trackFile:
+                trackData = trackFile.read()
+                if type == 'kml':
+                    trackData = self.minifyKml(trackData)
```
[the remark this answers is not in the document]
Comment (Sp3EdeR) [reply .github/scripts/build-tools/compact_data.py:58]: This is just data compacting. It is okay if it leaves the data as-is, if unsupported.

File: .github/scripts/build-tools/compact_data.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            print('Failed to optimize ' + route['md'], file=sys.stderr)
+
+    def importTrack(self, route):
+        type = ""
+        if 'kml' in route:
+            type = 'kml'
+        else:
+            return
+        if not route[type].endswith('.' + type):
+            return
```
[the remark this answers is not in the document]
Comment (Sp3EdeR) [reply .github/scripts/build-tools/compact_data.py:50]: I am no sure what I should be doing about this

File: .github/scripts/build-tools/compact_data.py (hunk 2 above)
Comment (Sp3EdeR) [reply .github/scripts/build-tools/compact_data.py:25]: Ah, good catch! It was supposed to be a for..else... The indentation got goopy.

File: .github/scripts/build-tools/compact_data.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nbali) [reply .github/scripts/build-tools/compact_data.py:50]: just fyi, it's strange to see as it's almost duplication

File: .github/scripts/build-tools/compact_data.py (hunk 2 above)
Comment (Sp3EdeR) [reply .github/scripts/build-tools/compact_data.py:25]: It works correctly. The example code would have to be run inside the for-loop to cover all values. Running intersection on all keys and route keys for EVERY key in a loop seems unnecessary. O(N^2) while mine is O(N)

**Review by nbali — APPROVED** [review]

**Pull request merged**